$OpenBSD: patch-objects_drawin_c,v 1.5 2014/10/13 09:00:48 dcoppa Exp $

commit f236a5f0c70015c0c2b3a57a8aec9dcc5dc49bdc
Author: Uli Schlachter <psychon@znc.in>
Date:   Sat Oct 11 13:27:29 2014 +0200

drawin_update_drawing: Remove optimization for invisible drawins

This fixes the following code:

   local d = drawin({})
   d.visible = true

The drawin now has a cairo surface assigned

   d.visible = false
   d.width = 1234
   d.visible = true

The width change while the drawin was not visible would not get propagated to
the drawable because of the code that this patch removes. The expectation was
that drawin_map() would update the drawable later.

However, because the drawin was already visible, its drawable also already has
a surface assigned. Thus, drawin_map() wouldn't update the drawable either.

Fix this by just removing this optimizations.

Signed-off-by: Uli Schlachter <psychon@znc.in>

commit 6673ecb167c5a5d85c8bacab15c1b6c0c7f65e80
Author: Uli Schlachter <psychon@znc.in>
Date:   Sat Oct 11 13:12:57 2014 +0200

drawin: Don't special-case moves

The code in drawin_moveresize() tries to be clever and only updates the drawing
state of the drawable when it is resized, not when it is moved around. This used
to be necessary because once upon a time, drawin_update_drawing() threw away all
of the drawing state and thus forcing a repaint. These days it just calls
drawable_set_geometry() as well and that function special-cases moves.

So this old code in drawin_moveresize() is no longer necessary and actually
caused problems.

These problems occurred because drawin_update_drawing() is being clever and
doesn't do anything for .visible = false drawins, because their drawing state
will be updated once they become visible. However, not skipping
drawable_set_geometry() means that this broke, because drawin_map() thought that
the drawing state was up to date while in reality it wasn't.

References: http://article.gmane.org/gmane.comp.window-managers.awesome/10852
Signed-off-by: Uli Schlachter <psychon@znc.in>

--- objects/drawin.c.orig	Fri Apr 11 11:07:10 2014
+++ objects/drawin.c	Mon Oct 13 10:39:02 2014
@@ -82,11 +82,6 @@ drawin_unref_simplified(drawin_t **item)
 static void
 drawin_update_drawing(drawin_t *w, int widx)
 {
-    /* If this drawin isn't visible, we don't need an up-to-date cairo surface
-     * for it. (drawin_map() will later make sure we are called again) */
-    if(!w->visible)
-        return;
-
     luaA_object_push_item(globalconf.L, widx, w->drawable);
     drawable_set_geometry(w->drawable, -1, w->geometry);
     lua_pop(globalconf.L, 1);
@@ -137,14 +132,7 @@ drawin_moveresize(lua_State *L, int udx, area_t geomet
         mask_vals |= XCB_CONFIG_WINDOW_HEIGHT;
     }
 
-    if(mask_vals & (XCB_CONFIG_WINDOW_WIDTH | XCB_CONFIG_WINDOW_HEIGHT))
-        drawin_update_drawing(w, udx);
-    else {
-        /* We still have to set x/y */
-        luaA_object_push_item(L, udx, w->drawable);
-        drawable_set_geometry(w->drawable, -1, w->geometry);
-        lua_pop(L, 1);
-    }
+    drawin_update_drawing(w, udx);
 
     /* Activate BMA */
     client_ignore_enterleave_events();
